llvm/include/llvm/Support/VirtualFileSystem.h

```diff
@@ -1125,6 +1125,54 @@ class YAMLVFSWriter {
   void write(llvm::raw_ostream &OS);
 };
 
+/// File system that tracks the number of calls to the underlying file system.
+/// This is particularly useful when wrapped around \c RealFileSystem to add
+/// lightweight tracking of expensive syscalls.
+class TracingFileSystem
+    : public llvm::RTTIExtends<TracingFileSystem, ProxyFileSystem> {
+public:
+  static const char ID;
+
+  std::size_t NumStatusCalls = 0;
```

Comment:
I'm contemplating whether to make these atomic. The root `FileSystem` class inherits from `ThreadSafeRefCountedBase`, suggesting that VFSs may be thread-safe.

For example in `clang-scan-deps`, all workers _could_ use single `RealFileSystem`. If we wrapped that in a single thread-safe `TracingFileSystem`, we'd very easily get the complete VFS stats. The current non-thread-safe approach would require having N of these VFS stacks and then manually aggregating the individual VFS stats.

Thoughts?

Reply:
Looking at the existing VFSs there appears to be a reasonable subset that currently are thread safe for stat/open/etc., but some definitely aren't. I think it's fine to make this one thread safe and document that it is if the underlying FS is.

I think they can also all be relaxed memory order, you have to have some other synchronization before collecting the numbers at the end anyway.

Reply:
@aganea Any thoughts? I'm asking because of https://github.com/llvm/llvm-project/pull/88427. I that making the counters atomic in this PR is not that problematic, since you can still choose how you use the VFS (i.e. how many per core you share), unlike the originally global counters in `FileManager` in your PR.

Reply:
I'm going to land this with non-atomic internals and add that feature when/if the need arises.